File: Robot Football/RobotFootball/NativeRouteFinders/AStarRouteFinder.cpp
#include "stdafx.h"

#include "GridSquare.h"
#include "PositionedObject.h"
#include "PointUtilities.h"

namespace NativeRouteFinders
{
	/// @brief Initializes a given grid using the information provided.
	void InitGrid(POINT startPoint, POINT endPoint, SIZE fieldSize, SIZE resolution, SIZE gridSize, 
				  GridSquare** grid, int objectClearance, SIZE movingObjectSize, 
				  PositionedObject* opponents, int opponentsCount);

	/// @brief Reconstructs a path from the given cameFrom information.
	unique_ptr<vector<POINT>> ReconstructPath(map<GridSquare*, GridSquare*>* cameFrom, GridSquare* currentSquare);
	/// @brief Calculates the heuristic score for the given square
	float CalculateHeuristic(GridSquare* square, POINT endPoint, bool useSqrt);
	/// @brief Calculates the length from one point to another
	float CalculateLength(POINT startPoint, POINT endPoint, bool useSqrt);

	/// @brief Initializes the Native A* Route Finder
	///
	/// This function is used by managed software to initialize the DLL for time-testing processes.
	/// It is used to ensure that the late-bound DLL has been loaded, so that the loading time does
	/// not register in the time tests.
	///
	/// @returns ERROR_SUCCESS under all conditions.
	NATIVEROUTEFINDERS_API HRESULT Init()
	{
		return ERROR_SUCCESS;
	}

	/// @brief Finds a route using the A* algorithm
	/// 
	/// @param startPoint The point to start the algorithm from.
	/// @param endPoint The point to find a route to
	/// @param fieldSize The size of the field that will be searched.
	/// @param resolution The size of each square in the field, once it has been discretized
	/// @param movingObjectSize The size of the object that will follow the route
	/// @param objectClearance The distance that obstacles must be cleared by
	/// @param opponents The opponents that act as obstacles
	/// @param opponentsCount The length of the opponents array
	/// @param routeResult The route that has been produced by the algorithm (an output parameter)
	/// @param routeResultLength The length of the routeResult array
	/// @param useSqrt If false, all distance algorithms skip the square root step of the calculation
	///
	/// @return ERROR_SUCCESS if a route has been found and placed in routeResult
	/// @return E_INVALIDARG if the specified resolution is 0
	/// @return E_FAIL if no route has been found
	/// @return ERROR_INSUFFICIENT_BUFFER if the find route is too long for the given buffer
	///
	/// @bug Does not currently detect when the 'neighboring squares have actually wrapped onto the next line
	///
	/// Determines a route from startPoint to endPoint using the A* algorithm.
	///
	/// Algorithm taken from @cite aiModernApproach.
	NATIVEROUTEFINDERS_API HRESULT AStarFindRoute(POINT startPoint, POINT endPoint, 
												  SIZE fieldSize, SIZE resolution, 
												  SIZE movingObjectSize, int objectClearance,
												  PositionedObject* opponents, int opponentsCount,
												  POINT* routeResult, unsigned int routeResultLength, bool useSqrt)
	{
		// Prevent any divide by zero errors
		if (resolution.cx < 1 || resolution.cy < 1)
			return E_INVALIDARG;

		// Calculate the number of squares in the grid
		SIZE gridSize;
		gridSize.cx = fieldSize.cx / resolution.cx;
		gridSize.cy = fieldSize.cy / resolution.cy;

		// Create the grid as an array (for speed of access)
		auto grid = new GridSquare*[gridSize.cx * gridSize.cy];

		InitGrid(startPoint, endPoint, fieldSize, resolution, gridSize, grid, objectClearance, movingObjectSize, opponents, opponentsCount);

		vector<GridSquare*> openSet;
		vector<GridSquare*> closedSet;
		map<GridSquare*, GridSquare*> cameFrom;

		// Would rather do this in parallel, but concurrent_vector doesn't allow you to remove things,
		// which makes later things more awkward.
		for_each(grid, grid + gridSize.cx * gridSize.cy, [=,&openSet](GridSquare* square)
		{
			if (square->Type != Origin)
				return;
			square->KnownScore = 0;
			square->HeuristicScore = CalculateHeuristic(square, endPoint, useSqrt);
			openSet.push_back(square);
		});

		// Loop through until there are no more squares to examine.
		while (!openSet.empty())
		{
			// Search the best squares first.
			auto smallestSquare = min_element(openSet.begin(), openSet.end(), [](GridSquare* square1, GridSquare* square2) { return square1->GetTotalScore() < square2->GetTotalScore(); });

			auto square = (*smallestSquare);
			if (square->Type == Destination)
			{
				// We've arrived!
				unique_ptr<vector<POINT>> path = ReconstructPath(&cameFrom, cameFrom[square]);
				path->push_back(square->Location);
				if (path->size() > routeResultLength)
					return ERROR_INSUFFICIENT_BUFFER;
				// Copy the result into the given buffer
				for (unsigned int i = 0; i < path->size(); i++)
				{
					routeResult[i].x = (*path)[i].x * resolution.cx;
					routeResult[i].y = (*path)[i].y * resolution.cy;
				}
				return ERROR_SUCCESS;
			}

			// Remove the square that has just been examined from the open set and put it into the closed set.
			openSet.erase(smallestSquare, smallestSquare+1);
			closedSet.push_back(square);

			auto index = IndexFromPoint(square->Location, gridSize.cx);

			int neighbourIndexes[] = { index + gridSize.cx + 1, 
									  index - gridSize.cx + 1,
									  index + gridSize.cx - 1, 
									  index - gridSize.cx - 1, index + 1, 
									  index - 1, 
									  index + gridSize.cx, 
									  index - gridSize.cx,
								  
			};


			// Check all of the current squares neighbours
			for (auto i = 0; i < 8; i++)
			{
				if (neighbourIndexes[i] < 0 || neighbourIndexes[i] >= gridSize.cx * gridSize.cy)
					continue; // Skip any indexes out of the range of the grid

				auto neighbour = grid[neighbourIndexes[i]];
			
				// Is the neighbour in the closed set?
				if (find(closedSet.begin(), closedSet.end(), neighbour) != closedSet.end())
					// Yes, so skip it as it's already been processed.
					continue;

				auto tentativeScore = square->KnownScore + CalculateLength(square->Location, neighbour->Location, useSqrt);
				auto tentativeIsBetter = false;

				// Is the neighbout in the open set?
				if (find(openSet.begin(), openSet.end(), neighbour) == openSet.end())
				{
					// No, so add it and make a guess at it's distance to the destination
					openSet.push_back(neighbour);
					neighbour->HeuristicScore = CalculateHeuristic(neighbour, endPoint, useSqrt);
					tentativeIsBetter = true;
				}
				else if (tentativeScore < neighbour->KnownScore)
					// Yes, and it's closer to the start than previously thought
					tentativeIsBetter = true;
				else
					// No
					tentativeIsBetter = false;

				// Is this square closer to the start than previously examined neigbouts?
				if (tentativeIsBetter)
				{
					// Mark this square as coming from the original square
					cameFrom[neighbour] = square;
					neighbour->KnownScore = tentativeScore;
				}
			}
		}

		// Run out of squares to examine without finding the destination, therefore no route possible.
		return E_FAIL;
	}
						
	/// @brief Finds a route using the A* algorithm
	/// 
	/// @param startPoint The point to start the algorithm from.
	/// @param endPoint The point to find a route to
	/// @param fieldSize The size of the field that will be searched.
	/// @param resolution The size of each square in the field, once it has been discretised
	/// @param movingObjectSize The size of the object that will follow the route
	/// @param objectClearance The distance that obstacles must be cleared by
	/// @param opponents The opponents that act as obstacles
	/// @param opponentsCount The length of the opponents array
	/// @param routeResult The route that has been produced by the algorithm (an output parameter)
	/// @param routeResultLength The length of the routeResult array
	/// @param useSqrt If false, all distance algorithms skip the square root step of the calculation
	///
	/// @return ERROR_SUCCESS if a route has been found and placed in routeResult
	/// @return E_INVALIDARG if the specified resultion is 0
	/// @return E_FAIL if no route has been found
	/// @return ERROR_INSUFFICIENT_BUFFER if the find route is too long for the given buffer
	///
	/// @bug Does not currently detect when the 'neighbouring' squares have actually wrapped onto the next line
	///
	/// Determines a route from startPoint to endPoint using the A* algorithm.  Uses an unordered_set to 
	/// store the open and closed sets to speed up searches.  Otherwise, the algorithm is identical to
	/// AStarFindRoute
	///
	/// Algorithm taken from @cite aiModernApproach.
	///
	/// @see AStarFindRoute
	NATIVEROUTEFINDERS_API HRESULT HashMapAStarFindRoute(POINT startPoint, POINT endPoint, 
												  SIZE fieldSize, SIZE resolution, 
												  SIZE movingObjectSize, int objectClearance,
												  PositionedObject* opponents, int opponentsCount,
												  POINT* routeResult, unsigned int routeResultLength, bool useSqrt)
	{
		// Prevent any divide by zero errors
		if (resolution.cx < 1 || resolution.cy < 1)
			return E_INVALIDARG;

		// Calculate the number of squares in the grid
		SIZE gridSize;
		gridSize.cx = fieldSize.cx / resolution.cx;
		gridSize.cy = fieldSize.cy / resolution.cy;

		// Create the grid as an array (for speed of access)
		auto grid = new GridSquare*[gridSize.cx * gridSize.cy];

		InitGrid(startPoint, endPoint, fieldSize, resolution, gridSize, grid, objectClearance, movingObjectSize, opponents, opponentsCount);

		unordered_set<GridSquare*> openSet;
		unordered_set<GridSquare*> closedSet;
		map<GridSquare*, GridSquare*> cameFrom;

		// Would rather do this in parallel, but concurrent_vector doesn't allow you to remove things,
		// which makes later things more awkward.
		for_each(grid, grid + gridSize.cx * gridSize.cy, [=,&openSet](GridSquare* square)
		{
			if (square->Type != Origin)
				return;
			square->KnownScore = 0;
			square->HeuristicScore = CalculateHeuristic(square, endPoint, useSqrt);
			openSet.insert(square);
		});

		while (!openSet.empty())
		{
			auto smallestSquare = min_element(openSet.begin(), openSet.end(), [](GridSquare* square1, GridSquare* square2) { return square1->GetTotalScore() < square2->GetTotalScore(); });

			auto square = (*smallestSquare);
			if (square->Type == Destination)
			{
				unique_ptr<vector<POINT>> path = ReconstructPath(&cameFrom, cameFrom[square]);
				path->push_back(square->Location);
				if (path->size() > routeResultLength)
					return ERROR_INSUFFICIENT_BUFFER;
				for (unsigned int i = 0; i < path->size(); i++)
				{
					routeResult[i].x = (*path)[i].x * resolution.cx;
					routeResult[i].y = (*path)[i].y * resolution.cy;
				}
				return ERROR_SUCCESS;
			}

			openSet.erase(smallestSquare);
			closedSet.insert(square);

			auto index = IndexFromPoint(square->Location, gridSize.cx);

			int neighbourIndexes[] = { index + gridSize.cx + 1, 
									  index - gridSize.cx + 1,
									  index + gridSize.cx - 1, 
									  index - gridSize.cx - 1, index + 1, 
									  index - 1, 
									  index + gridSize.cx, 
									  index - gridSize.cx,
								  
			};

			for (auto i = 0; i < 8; i++)
			{
				if (neighbourIndexes[i] < 0 || neighbourIndexes[i] >= gridSize.cx * gridSize.cy)
					continue; // Skip any indexes out of the range of the grid

				auto neighbour = grid[neighbourIndexes[i]];
			
				if (closedSet.find(neighbour) != closedSet.end())
					continue;

				auto tentativeScore = square->KnownScore + CalculateLength(square->Location, neighbour->Location, useSqrt);
				auto tentativeIsBetter = false;

				if (openSet.find(neighbour) == openSet.end())
				{
					openSet.insert(neighbour);
					neighbour->HeuristicScore = CalculateHeuristic(neighbour, endPoint, useSqrt);
					tentativeIsBetter = true;
				}
				else if (tentativeScore < neighbour->KnownScore)
					tentativeIsBetter = true;
				else
					tentativeIsBetter = false;

				if (tentativeIsBetter)
				{
					cameFrom[neighbour] = square;
					neighbour->KnownScore = tentativeScore;
				}
			}
		}

		return E_FAIL;
	}

    /// Takes the already allocated GridSquare array grid and sets it up for use by AStarFindRoute and 
	/// HashMapAStarFindRoute.
	///
	/// Uses objectClearance and movingObjectSize to determine the apparent size of the opponents,
	/// and marks the squares that contain them as SquareType::Obstacle.
	///
	/// Marks the square that contains endPoint as SquareType::Destination.
	///
	/// Marks the square that contains startPoint as SquareType::Origin.
	///
	/// Works in parallel as far as possible, using the Microsoft Parallel Patterns Library 
	/// (http://msdn.microsoft.com/en-us/library/dd492418.aspx)
	void InitGrid(POINT startPoint, POINT endPoint, SIZE fieldSize, SIZE resolution, SIZE gridSize, 
				  GridSquare** grid, int objectClearance, SIZE movingObjectSize, 
				  PositionedObject* opponents, int opponentsCount)
	{
		// The number of elements in the grid array
		const int gridLength = gridSize.cx * gridSize.cy;

		// Work out the size of the player in squares
		SIZE playerSize;
		playerSize.cx = (LONG)ceil((float)(movingObjectSize.cx + objectClearance) / (resolution.cx * objectClearance));
		playerSize.cy = (LONG)ceil((float)(movingObjectSize.cy + objectClearance) / (resolution.cy * objectClearance));

		// Make the clearance around the players square to simplify things
		auto clearance = max(playerSize.cx,playerSize.cy);

		// Initialise each array element and set it's GridSquare::Location based on it's array location.
		parallel_for(0, gridLength, [=](int i) { grid[i] = new GridSquare(); grid[i]->Location = PointFromIndex(i, gridSize.cx); });

		// Set up the area covered by each opponent as a SquareType::Obstacle
		parallel_for(0, opponentsCount, [=](int i) { 
			auto opponent = opponents[i];

			POINT centerGridPoint;
			centerGridPoint.x = (LONG)(floorf((float)opponent.Position.x / resolution.cx));
			centerGridPoint.y = (LONG)(floorf((float)opponent.Position.y / resolution.cy));

			auto minX = max(0, centerGridPoint.x - playerSize.cx - clearance);
			auto maxX = min(centerGridPoint.x + playerSize.cx + clearance, gridSize.cx);
			auto minY = max(0, centerGridPoint.y - playerSize.cy - clearance);
			auto maxY = min(centerGridPoint.y + playerSize.cy + clearance, gridSize.cy);

			for (auto i = minX; i < maxX; i++)
			{
				for (auto j = minY; j < maxY; j++)
				{
					grid[IndexFromPoint(i, j, gridSize.cx)]->Type = Obstacle;
				}
			}
		});

		// Mark the destination and origin of the route	
		auto gridEndX = (LONG)floorf((float)endPoint.x / resolution.cx);
		auto gridEndY = (LONG)floorf((float)endPoint.y / resolution.cy);

		grid[IndexFromPoint(gridEndX, gridEndY, gridSize.cx)]->Type = Destination;

		auto gridStartX = (LONG)floorf((float)startPoint.x / resolution.cx);
		auto gridStartY = (LONG)floorf((float)startPoint.y / resolution.cy);

		grid[IndexFromPoint(gridStartX, gridStartY, gridSize.cx)]->Type = Origin;
	}

	/// @param startPoint The point to measure from
	/// @param endPoint The point to measure to
	/// @param useSqrt If set to false, skip the square-root step
	/// 
	/// Calculates the distance from startPoint to endPoint using the Euclidian distance.
	///
	/// @note Setting useSqrt to false does not appear to produce any time savings.  In fact, 
	/// some experimentation suggests it takes longer
	/// @return A floating point number representing the distance from startPoint to endPoint
	float CalculateLength(POINT startPoint, POINT endPoint, bool useSqrt)
	{
		auto xLength = startPoint.x - endPoint.x;
		auto yLength = startPoint.y - endPoint.y;

		double val = xLength*xLength + yLength*yLength;

		if (useSqrt)
			val = sqrt(val);
		return (float)val;
	}

	/// @param square The square to calculate the heuristic for
	/// @param endPoint The current point to aim for
	/// @param useSqrt If false, all distance algorithms skip the square root step of the calculation
	///
	/// Calculates the heuristic score for the given square.
	///
	/// @return \f$+\infty\f$ if the square is an obstacle
	/// @return CalculateLength(square->Location, endPoint, useSqrt) otherwise
	float CalculateHeuristic(GridSquare* square, POINT endPoint, bool useSqrt)
	{
		if (square->Type == Obstacle)
			return numeric_limits<float>::infinity();
		return CalculateLength(square->Location, endPoint, useSqrt);
	}

	/// @param cameFrom The list of 'came from' details produced by the A* algorithm
	/// @param currentSquare The current square to examine
	///
	/// A recursive algorithm that reconstructs the path that has been produced by the A* algorithm
	///
	/// Algorithm taken from @cite aiModernApproach
	unique_ptr<vector<POINT>> ReconstructPath(map<GridSquare*, GridSquare*>* cameFrom, GridSquare* currentSquare)
	{
		if (cameFrom->find(currentSquare) != cameFrom->end())
		{
			auto path = ReconstructPath(cameFrom, cameFrom->at(currentSquare));
			path->push_back(currentSquare->Location);
			return path;
		}
		unique_ptr<vector<POINT>> path(new vector<POINT>());
		path->push_back(currentSquare->Location);
		return path;
	}
}

File: Robot Football/RobotFootball/AStarRouteFinder/AStarRouteFinder.cs
using System;
using System.Drawing;
using System.Linq;
using RobotFootballCore.Interfaces;
using RobotFootballCore.Objects;
using RobotFootballCore.RouteObjects;
using RobotFootballCore.Utilities;
using System.Threading.Tasks;
using System.Collections.Generic;

namespace RouteFinders
{
    /// <summary>
    /// A route finder that determines a route using the A* algorithm
    /// </summary>
    public class AStarRouteFinder : RouteFinder
    {
        /// <summary>
        /// The distance that objects must be cleared by.
        /// </summary>
        public int ObjectClearance { get; set; }

        /// <summary>
        /// Initialises a new instance of the AStarRouteFinder class.
        /// </summary>
        /// 
        /// Sets all member values to their defaults of:
        /// 
        /// * Resolution = 10 x 10
        /// 
        /// * ObjectClearance = 20
        public AStarRouteFinder()
        {
            Resolution = new Size(10, 10);
            ObjectClearance = 20;
        }

        /// <summary>
        /// Initialises a discrete grid using the information provided.
        /// </summary>
        /// <param name="startPoint">The point to start the route at</param>
        /// <param name="endPoint">The point to end the route at</param>
        /// <param name="field">The field that needs to be discretised</param>
        /// <param name="gridSize">The size of the grid to be palced placed over <paramref name="field"/></param>
        /// <param name="movingObject">The object that will be moving.</param>
        /// <returns>A GridSquare array, initialised to represent <paramref name="field"/>.</returns>
        /// 
        /// Produces a GridSquare array that is set up for use by <see cref="FindPath"/>.
        /// 
        /// Uses the ObjectClearance and <paramref name="movingObject" />'s <see cref="IPositionedObject::Size"/> property to determine the 
        /// apparent size of opponents, and marks the squares that contain them as <see cref="SquareType::Obstacle"/>.
        /// 
        /// Marks the square that contains <paramref name="endPoint" /> as <see cref="SquareType::Destination"/>.
        ///
        /// Marks the square that contains <paramref name="startPoint" /> as <see cref="SquareType::Origin"/>.
        /// 
        /// Works in parallel as far as possible, using the Microsoft Task Parallel Library (http://msdn.microsoft.com/en-us/library/dd460717.aspx).
        protected GridSquare[] InitGrid(PointF startPoint, PointF endPoint, Field field, Size gridSize, IPositionedObject movingObject)
        {
            var playersize = (movingObject.Size + new Size(ObjectClearance, ObjectClearance)).Scale(Resolution).Scale(2.0f).Ceiling();
            var clearance = Math.Max(playersize.Width, playersize.Height); // The amount to increase an obstacle's size by to allow for the player's size

            var grid = new GridSquare[gridSize.Height * gridSize.Width];
            // Initialize the grid
            Parallel.For(0, grid.Length, i => { 
                grid[i] = new GridSquare();
                grid[i].Location = PointExtensions.FromIndex(i, gridSize.Width); 
            });

            Parallel.ForEach(from p in field.Players where p.Team == Team.Opposition select p, player =>
            {
                var centerGridPoint = player.Position.Scale(Resolution).Floor();
                
                var minX = Math.Max(0, centerGridPoint.X - playersize.Width - clearance);
                var maxX = Math.Min(centerGridPoint.X + playersize.Width + clearance, gridSize.Width);
                var minY = Math.Max(0, centerGridPoint.Y - playersize.Height - clearance);
                var maxY = Math.Min(centerGridPoint.Y + playersize.Height + clearance, gridSize.Height);

                for (var i = minX; i < maxX; i++)
                {
                    for (var j = minY; j < maxY; j++)
                    {
                        if (i < 0 || j < 0)
                            continue;

                        var gridPoint = new Point(i, j);
                        grid[gridPoint.ToIndex(gridSize.Width)].Type = SquareType.Obstacle;
                    }
                }
            });

            var gridEndPoint = endPoint.Scale(Resolution).Floor();

            grid[gridEndPoint.ToIndex(gridSize.Width)].Type = SquareType.Destination;

            var gridStartPoint = startPoint.Scale(Resolution).Floor();
            grid[gridStartPoint.ToIndex(gridSize.Width)].Type = SquareType.Origin;

            return grid;
        }

        /// <summary>
        /// Finds a route using the A* algorithm
        /// </summary>
        /// <param name="startPoint">The point to start the algorithm from.</param>
        /// <param name="endPoint">The point to find a route to.</param>
        /// <param name="field">The field in which the route must be found.</param>
        /// <param name="movingObject">The object that will move around the <paramref name="field"/>.</param>
        /// <returns>
        /// A Route if a route can be found.
        /// 
        /// null otherwise
        /// </returns>
        /// <exception cref="InvalidOperationException">
        /// Thrown if the specified resolution is 0 in one or more directions.
        /// </exception>
        /// 
        /// Determines a route from <paramref name="startPoint" /> to <paramref name="endPoint" /> using the A* algorithm.
        ///
        /// Algorithm taken from @cite aiModernApproach
        public override Route FindPath(PointF startPoint, PointF endPoint, Field field, IPositionedObject movingObject)
        {
            if (Resolution.Height < 1 || Resolution.Width < 1)
                throw new InvalidOperationException("Resolution must be greater than or equal to 1 pixel by 1 pixel");

            var gridSize = field.Size.Scale(Resolution).Ceiling();

            var grid = InitGrid(startPoint, endPoint, field, gridSize, movingObject);

            var startPoints = from g in grid where g.Type == SquareType.Origin select g;

            var closedSet = new List<GridSquare>(); // The already checked points
            var openSet = new List<GridSquare>(from g in grid where g.Type == SquareType.Origin select g); // The points to check
            var cameFrom = new Dictionary<GridSquare, GridSquare>(); // A list of route data already calculated

            // Initialise the origin points
            Parallel.ForEach(startPoints, g => { g.KnownScore = 0; g.HeuristicScore = CalculateHeuristic(g, endPoint); });

            // While there are points available to check
            while (openSet.Any())
            {
                var square = openSet.OrderBy(p => p.TotalScore).First();
                if (square.Type == SquareType.Destination)
                {
                    var path = ReconstructPath(cameFrom, cameFrom[square]);
                    path.Path.Add(new LineSegment(cameFrom[square].Location, square.Location));
                    path.Scale(Resolution.Invert()); // Convert the path back to world coordinates from grid coordinates
                    return path;
                }

                openSet.Remove(square);
                closedSet.Add(square);

                var index = square.Location.ToIndex(gridSize.Width);

                // Calculate the neighbouring indexes and discard the ones out of range.
                var neighbourIndexes = new[] { index + 1, 
                                               index - 1, 
                                               index + gridSize.Width, 
                                               index - gridSize.Width,
                                               index + gridSize.Width + 1, 
                                               index - gridSize.Width + 1,
                                               index + gridSize.Width - 1, 
                                               index - gridSize.Width - 1 }.Where(i => (i >= 0) && (i < grid.Length));

                foreach (var i in neighbourIndexes)
                {
                    var neighbour = grid[i];

                    if (closedSet.Contains(neighbour))
                        continue; // Already checked this square. Skip it

                    // Work out the distance to the origin
                    var tentativeKnownScore = square.KnownScore + CalculateLength(square.Location, neighbour.Location);

                    bool tentativeIsBetter = false;

                    if (!openSet.Contains(neighbour))
                    {
                        // First time this point has been considered. The current distance is the best guess.
                        openSet.Add(neighbour);
                        neighbour.HeuristicScore = CalculateHeuristic(neighbour, endPoint);
                        tentativeIsBetter = true;
                    }
                    else if (tentativeKnownScore < neighbour.KnownScore)
                        // Point has been considered before and the last consideration was given a higher score, so use the knew one.
                        tentativeIsBetter = true;
                    else
                        // Point has been considered before and the last consideration was given a lower score, so use that one.
                        tentativeIsBetter = false;

                    if (tentativeIsBetter)
                    {
                        // If necessary, update the square's known score and note the path to that square.
                        cameFrom[neighbour] = square;
                        neighbour.KnownScore = tentativeKnownScore;
                    }
                }
            }

            return null;
        }

        /// <summary>
        /// Reconstructs a path from the given cameFrom information.
        /// </summary>
        /// <param name="cameFrom">The list of 'came from' details produced by the A* algorithm</param>
        /// <param name="currentSquare">The current square to examine</param>
        /// <returns>The shortest Route to the destination</returns>
        /// 
        /// A recursive algorithm that reconstructs the path that has been produced by the A* algorithm
        protected Route ReconstructPath(Dictionary<GridSquare, GridSquare> cameFrom, GridSquare currentSquare)
        {
            if (cameFrom.ContainsKey(currentSquare))
            {
                var path = ReconstructPath(cameFrom, cameFrom[currentSquare]);
                path.Path.Add(new LineSegment(cameFrom[currentSquare].Location, currentSquare.Location));
                return path;
            }
            return new Route();
        }

        /// <summary>
        /// Calculates the heuristic score for the given square.
        /// </summary>
        /// <param name="square">The square to calculate the heuristic for </param>
        /// <param name="endPoint">The current point to aim for </param>
        /// <returns>
        /// \f$+\infty\f$ if the square is an obstacle
        /// 
        /// CalculateLength(square.Location, endPoint) otherwise
        /// </returns>
        protected float CalculateHeuristic(GridSquare square, PointF endPoint)
        {
            if (square.Type == SquareType.Obstacle)
                return float.PositiveInfinity;
            else
                return CalculateLength(square.Location, endPoint);
        }

        
    }
}
